homeassistant/components/group/sensor.py

```diff
@@ -665,19 +675,40 @@ def _get_valid_units(self) -> set[str | None]:
 
         If device class is set and compatible unit of measurements.
         If device class is not set, use one unit of measurement.
+        Only calculate valid units if there are no valid units set.
         """
-        if (
-            device_class := self.device_class
-        ) in UNIT_CONVERTERS and self.native_unit_of_measurement:
+        # If we once have a set of valid units we should not recalculate it
+        # when all states needs to be considered
+        if (valid_units := self._valid_units) and not self._ignore_non_numeric:
+            return valid_units
+
+        native_uom = self.native_unit_of_measurement
+        if (device_class := self.device_class) in UNIT_CONVERTERS and native_uom:
             self._can_convert = True
             return UNIT_CONVERTERS[device_class].VALID_UNITS
-        if (
-            device_class
-            and (device_class) in DEVICE_CLASS_UNITS
-            and self.native_unit_of_measurement
-        ):
+        if device_class and (device_class) in DEVICE_CLASS_UNITS and native_uom:
             valid_uoms: set = DEVICE_CLASS_UNITS[device_class]
             return valid_uoms
-        if device_class is None and self.native_unit_of_measurement:
-            return {self.native_unit_of_measurement}
+        if device_class is None and native_uom:
+            return {native_uom}
         return set()
+
+    def _get_state_of_entities(
+        self,
+    ) -> list[str]:
+        """Return if states are valid."""
```

Comment:
The name of the method and the docstring don't agree. The method returns a list of entity_ids which have a valid state, that should be reflected in both the method name and docstring.